include/ftxui/screen/screen.hpp

```diff
@@ -18,9 +18,16 @@ using Element = std::shared_ptr<Node>;
 /// @brief A unicode character and its associated style.
 /// @ingroup screen
 struct Pixel {
+  // The graphemes stored into the pixel. To support combining characters,
+  // like: a⃦, this can potentially contains multiple codepoitns.
+  // Required: character.size() >= 1;
+  std::wstring character = L" ";
```

Comment:
Could you explain what kind of performance killer do you think about? Memory or CPU?

My reasoning was C++ implementations of std::string are using small objects allocations to avoid doing allocations on the heap. Small object is exactly what I would have 99..99% of the time. A fixed array is the same as std::string + small-object-optimization, but without the flexibility.
However, yes we can still compare `sizeof(std::wstring)` vs `sizeof(wchar_t[N]) ` and give the fixed array the avantage if N is small enough.

https://shaharmike.com/cpp/std-string/#:~:text=Simply%20put%2C%20Small%20Object%20Optimization,fields%20of%20std%3A%3Astring%20.&text=With%20a%20size%20of%2040,dedicated%20to%20hold%20small%20strings.

BTW: Do you know know if there is a limit of codepoints/grapheme that could be used for the fixed array?

Reply:
> also note that if you use UTF-8 instead of UTF-32 you can store more codepoints in less space in average).

That's a really good idea! I will try it.

Reply:
For posterity, I tried using UTF-8 instead of UTF-32, but
sizeof(std::string) == sizeof(std::wstring) = 32. In both case, no allocations are made, thank to small string optimization.
So there is no real improvement swiching from one to the other.

https://github.com/ArthurSonzogni/FTXUI/commit/609870c955f703a374711d0e53e6b23c67f43e26

I am not sure I am going to use, given it doesn't lead to any improvements.

Reply:
The reason why I said strings kill performance is because that's what happened to me when I tried to do this in Turbo Vision. However, Turbo Vision is designed in such a way that screen cell buffers are copied around very often and thus it is important to optimize for copy operations. For that reason I chose to use a fixed-length string that could be `memcpy`'d. FXTUI works differently and it even looks like the text in a screen cell is often manipulated directly, so it may be more convenient to just use a regular string.

> Do you know know if there is a limit of codepoints/grapheme that could be used for the fixed array?

There is no such limit. The basis for using a fixed-length string is that natural language rarely uses more than a few consecutive combining characters. Use cases potentially requiring large amounts of combining characters are emojis (e.g. 👩‍❤️‍💋‍👨) and Zalgo text (e.g. Z̸̪̉â̸̪l̶͔̇ǵ̸͇ò̸̧ ̵͂ͅt̶͍͗ë̸̻́x̸̮͑t̶͍́).

> For posterity, I tried using UTF-8 instead of UTF-32, but sizeof(std::string) == sizeof(std::wstring) = 32.

The basis for preferring UTF-8 is that combining characters are often just 2 or 3 bytes long. Therefore, in a fixed-length string you can fit more characters if using UTF-8 than if using UTF-32.

In Turbo Vision I chose to reserve only 12 bytes of text per cell, so UTF-8 turned out very convenient. However, I was already using UTF-8 all across the API, which is not the case of FTXUI. So, as you say, there are no apparent advantages to replacing `std::wstring` with `std::string`.

> My reasoning was C++ implementations of std::string are using small objects allocations to avoid doing allocations on the heap.

Despite this optimization, `std::string` entails a considerable overhead compared to a fixed-length array:
* It is not trivially copyable (`memcpy`-able).
* Operations like copy, comparison, etc. require checking whether the small string optimization is being used or not; some implementations of the STL also make this check when accessing the string data or length.
* Operations like assignment and concatenation also perform other checks such as whether an exception needs to be raised.

Yet, there is no reason for you to stop using the STL strings if you don't observe a performance issue.